lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosAlgorithm.cxx

```diff
@@ -268,6 +253,39 @@ AdaptiveStrategy FunctionalChaosAlgorithm::getAdaptiveStrategy() const
   return adaptiveStrategy_;
 }
 
+Distribution FunctionalChaosAlgorithm::BuildDistribution(const Sample & inputSample)
+{
+  // Recover the distribution, taking into account that we look for performance
+  // so we avoid to rebuild expensive distributions as much as possible
+  const UnsignedInteger inputDimension = inputSample.getDimension();
+  Collection< Distribution > marginals(inputDimension);
+  // The strategy is to test first standard distributions, else fallback to HistogramFactory
+  // Beta, Gamma yield slower iso transformations than HistogramFactory
+  // in the other hand HistogramFactory transformation shrinks the range and may results in a metamodel with saturations
+  Collection< DistributionFactory > factories;
+  factories.add(NormalFactory());
+  factories.add(UniformFactory());
+  factories.add(GammaFactory());
+  factories.add(BetaFactory());
+  LOGINFO("In FunctionalChaosAlgorithm, identify marginal distributions");
+  const Description inputDescription(inputSample.getDescription());
+  for (UnsignedInteger i = 0; i < inputDimension; ++ i)
+  {
+    TestResult bestResult;
+    // Here we remove the duplicate entries in the marginal sample using the automatic compaction of the support in the UserDefined class
+    const Sample marginalSample(inputSample.getMarginal(i).sortUnique());
+    const Distribution candidate(FittingTest::BestModelKolmogorov(marginalSample, factories, bestResult));
+    // This threshold is somewhat arbitrary. It is here to avoid expensive kernel smoothing.
+    if (bestResult.getPValue() > ResourceMap::GetAsScalar("FunctionalChaosAlgorithm-PValueThreshold")) marginals[i] = candidate;
+    else marginals[i] = HistogramFactory().build(marginalSample);
```

Comment:
we could set Histogram or KernelSmoothing. Choice could be oriented using a ResouceMap key